### DISCUSSION TITLE
OWASP HTTP Server Header - How to Disable or Non-Display Informative Values

### DISCUSSION THREAD

**Discussion by devcoisa** [DefectDojo/django-DefectDojo#9793] [Q&A]
Hello,
to address an internal security audit we should not display the Server header or at least not showing informative values, following the recommendation of OWASP HTTP Header Cheat Sheet: https://cheatsheetseries.owasp.org/cheatsheets/HTTP_Headers_Cheat_Sheet.html#server

OWASP HTTP Header Cheat Sheet recommendation is to remove this header or set non-informative values. Did you have anything similar like [this](https://github.com/DefectDojo/django-DefectDojo/issues/5328) to address this potential vulnerability i am asking for ? 

As most likely we don't have any DD initiative to set this header to OFF, is it possible to customise it with non-informative values ? 

Thanks

**Comment by mtesauro** [comment]
You don't mention how you've deployed DefectDojo so I'll have to answer generally:

For the most deploys of DefectDojo, nginx is used to front the Django app (e.g. the nginx container in docker compose)

You can customize the nginx.conf file to set the server_token to off - more information is available from the Nginx docs at https://nginx.org/en/docs/http/ngx_http_core_module.html#server_tokens

Where and how to edit the nginx.conf depends on the deploy used.

**Comment by devcoisa** [comment]
Hi,
we already have on the nginx conf (configmap) the server tokens directive set to off
      server {
        listen 8080;
        location = /50x.html {
          root /usr/share/nginx/html;
        server_tokens off;
        }

, although but we are still getting the nginx version on the response http header field.

**Comment by devcoisa** [comment]
Hi @mtesauro,
are you aware of any limitation with the customization of server response headers ?
We've seen to customize server token headers we should have commercial license: https://nginx.org/en/docs/http/ngx_http_core_module.html#server_tokens

And we've already tested on our env the server_tokens directive to off either the http section and doesn't work.

Ideas ?

**Comment by mtesauro** [comment]
I don't know how you have it installed/configured but I either bind-mount in a nginx.config so I can customize it as I want or need or docker build my own version of the Nginx container with my custom nginx.conf included in that build.

An example of bind-mounting Nginx conf into the container is at https://github.com/DefectDojo/Community-Contribs/blob/master/CentOS-7-deploy/dojo/docker-compose.yml#L19

**Reply by devcoisa** [reply]
Hi @mtesauro,
our deployment model is based on applying "kubectl apply -k" of a specific subset of directories containing the manifest files (yaml) with the base/overlay/conf files (nginx, rabitmq,etc...).
The example below is a code snippet with the nginx base configmap we are deploying. As you can see we're applying the server_tokens directive on server location but is not disabling emitting nginx version the “Server” response header field., showing up the following info "Server: nginx/1.25.4".

![image](https://github.com/DefectDojo/django-DefectDojo/assets/160020684/e77abf76-edd1-4eb3-955b-0896cbd4ded2)


_apiVersion: v1
kind: ConfigMap
metadata:
  namespace: '@@KUBE_NameSpace@@'
  name: '@@KUBE_NameSpace@@-defectdojo-nginx-configmap'
data:
  nginx.conf: |
    worker_processes 1;
    error_log /var/log/nginx/error.log warn;
    pid /var/run/defectdojo/nginx.pid;
    events {
      worker_connections  1024;
    }
    http {
      include /etc/nginx/mime.types;
      default_type application/octet-stream;
      map $http_x_forwarded_for $client_ip {"~(?<IP>([0-9]{1,3}\.){3}[0-9]{1,3}).*" $IP;}
      log_format  main  '$remote_addr - $remote_user [$time_local] "$request" '
                        '$status $body_bytes_sent "$http_referer" '
                        '"$http_user_agent" "$http_x_forwarded_for"';
      access_log /var/log/nginx/access.log main;
      client_max_body_size 800m;
      sendfile on;
      keepalive_timeout 65;
      upstream uwsgi_server {
        include /run/defectdojo/uwsgi_server;
      }
      server {
        listen 8080;
        location = /50x.html {
          root /usr/share/nginx/html;
        **server_tokens off;**
        }
        location /static {
          alias /usr/share/nginx/html/static;
        }
        location /media {
            alias /usr/share/nginx/html/media;
        }
        location / {
          include /run/defectdojo/uwsgi_pass;
          include /etc/nginx/wsgi_params;
          uwsgi_read_timeout 1800;
          uwsgi_param HTTP_X_FORWARDED_FOR $client_ip;
        }
        location /django_metrics {
          # do no edit the following lines, instead set the environment  
          # variables METRICS_HTTP_AUTH_USER and METRICS_HTTP_AUTH_PASSWORD
          #auth_basic "Metrics";
          #auth_basic_user_file /etc/nginx/.htpasswd;
          include /run/defectdojo/uwsgi_pass;
          include /etc/nginx/wsgi_params;
          uwsgi_read_timeout 1800;
        }
        location /nginx_status {
          # do no edit the following lines, instead set the environment  
          # variables METRICS_HTTP_AUTH_USER and METRICS_HTTP_AUTH_PASSWORD
          #auth_basic "Metrics";
          #auth_basic_user_file /etc/nginx/.htpasswd;
          #stub_status  on;
          access_log   off;
        }
        # Used by Kubernetes liveness checks
        location = /nginx_health {
          return 200 "Born to be alive!\n";
          access_log off;
        }
        # Used by Kubernetes readiness checks
        location = /uwsgi_health {
          limit_except GET { deny all; }
          rewrite /.+ /login?next=/ break;
          include /run/defectdojo/uwsgi_pass;
          include /etc/nginx/wsgi_params;
          access_log off;
        }
        error_page 500 502 503 504 /50x.html;
      }
    }_
